Gaurav recipe app

gauravThakur-sgl opened the pull request:
# Recipe App
### The app enables users to discover recipes through ingredient-based search and dietary preferences, leveraging TheMealDB API for recipe data.

## Key Features
### Recipe Search & Discovery

- Keyword -based recipe search
- Real-time search results updates
- Filter for selecting areas and Categories

### Recipe Details

- Comprehensive recipe information display
- Step-by-step cooking instructions
- Ingredient lists with measurements

### User Features

- Recipe favoriting functionality
- Persistent storage of saved recipes

## Api Integration

- Recipe Search: `/search.php?s=`
- letter-based-search `/search.php?f=`
- Category filtering: `/filter.php?c=`
- Ingredient filtering: /`filter.php?i=`

## Pages

### 1.  HomePage

- Search Functionality
- Save or remove recipe
- Featured recipes display
- filter
- Saved Recipe Button
- Pagination
- Reset Button

Screen Shot
![HomePage](https://github.com/user-attachments/assets/97699998-7108-4d49-bf81-d4105252d4e5)




### 2. Search Results

-  Results based on grid
- Pagination
- Save or remove recipe

Screen Shot
![SearchResults](https://github.com/user-attachments/assets/c82faca7-ff49-4d57-b353-3dc8034a2724)


### 3. Recipe Details

- Recipe Information Display
- Instruction View
- Youtube video
- Ingredients and its usage

Recipe Details
![Recipe Details](https://github.com/user-attachments/assets/653f1dbd-8a38-446f-83c9-e2f1b149f5fe)


### 4. Saved Recipe
- Saved Recipe Cards
- Click click to view the recipe

ScreenShot 
![savedRecipe](https://github.com/user-attachments/assets/b7833fbe-b5ec-432f-a93e-3888382299ea)

shubhamships commented on gaurav/src/App.tsx:
```diff
@@ -1,53 +1,20 @@
+    <BrowserRouter>
+      <Routes>
+        <Route path="/" element={<Root />} />
+        <Route path="searchresults" element={<SearchResults />} />
```
`path="search-results"`

shubhamships commented on gaurav/src/App.tsx:
```diff
@@ -1,53 +1,20 @@
+      <Routes>
+        <Route path="/" element={<Root />} />
+        <Route path="searchresults" element={<SearchResults />} />
+        <Route path="recipedetail/:id" element={<RecipeDetail />} />
```
`path="recipe-detail"`

shubhamships commented on gaurav/src/app/store.ts:
```diff
@@ -1,18 +1,18 @@
+// import { combineReducers, configureStore } from "@reduxjs/toolkit";
+// import punchDataReducer from "../redux/reducers.ts"
+// import { persistReducer, persistStore } from "redux-persist";
+// import storage from "redux-persist/lib/storage";
```
remove commented code

shubhamships commented on gaurav/src/main.tsx:
```diff
@@ -10,12 +10,12 @@ import { InstallPWA } from "./components/InstallPWA.tsx";
-    <Provider store={store}>
-      <PersistGate persistor={persistor}>
+    {/* <Provider store={store}> */}
+      {/* <PersistGate persistor={persistor}> */}
```
remove commented code

shubhamships commented on gaurav/src/pages/recipe/HomePage.tsx:
```diff
@@ -0,0 +1,262 @@
+
+  const fetchRecipes = async (query = "", category = "all", area = "all") => {
+    try {
+      let url = `https://www.themealdb.com/api/json/v1/1/search.php?s=${query}`;
```
make a reusable api url constant variable and use it here

shubhamships commented on gaurav/src/pages/recipe/HomePage.tsx:
```diff
@@ -0,0 +1,262 @@
+
+  const fetchAreas = async () => {
+    try {
+      const res = await axios.get("https://www.themealdb.com/api/json/v1/1/list.php?a=list");
```
use the API_URL variable here

shubhamships commented on gaurav/src/pages/recipe/SavedRecipe.tsx:
```diff
@@ -0,0 +1,94 @@
+  strMealThumb: string;
+  strTags: string;
+  strYoutube: string;
+  strInstructions: string;
```
same interface is being used above, 

declare and export it in a separate interfaces.ts file inside lib

shubhamships commented on gaurav/src/pages/recipe/SearchResults.tsx:
```diff
@@ -0,0 +1,173 @@
+
+  const fetchRecipes = async (query = ""): Promise<IData[]> => {
+    try {
+      const res = await axios.get(`https://www.themealdb.com/api/json/v1/1/search.php?s=${query}`);
```
use URL variable

shubhamships commented on gaurav/src/pages/recipe/SearchResults.tsx:
```diff
@@ -0,0 +1,173 @@
+
+  return (
+    <>
+      <div className="bg-[#083344] min-h-screen relative">
```
bg-[#083344] - adding color codes like this is bad practice. Use CSS variable names

shubhamships commented on gaurav/src/pages/recipe/components/CardDetail.tsx:
```diff
@@ -0,0 +1,71 @@
+import { MapPin, Tags, Utensils } from "lucide-react";
+import { CardSmallDetail } from "./CardSmallDetail";
+interface IRecipeData {
```
Make one resusable RecipeData Interface and extend it using inheritance

shubhamships commented on gaurav/src/routes/Root.tsx:
```diff
@@ -1,29 +1,79 @@
+  const handleReset = () => {
+    try {
+      localStorage.clear();
+      window.location.reload();
```
use location.reload where `location = useLocation()` instead of `window.location.reload()`

shubhamships commented on gaurav/tailwind.config.js:
```diff
@@ -57,7 +60,7 @@ export default {
       },
     },
-    plugins: [tailwindcssAnimate],
+    // plugins: [tailwindcssAnimate],
```
remove commented code from everywhere

gauravThakur-sgl commented:
**@shubhamships Sir**
Updated Changes :- 
- use `API_URL` variable for reusable api link
- Removed Commented Code from everywhere
- added `interfaces.ts` in `lib` for Reusable interface Class
- Removed custom colors from `tailwindcss.config` file
- use `location.reload` instead of window.location.reload()

shubhamships commented on gaurav/src/pages/recipe/HomePage.tsx:
```diff
@@ -0,0 +1,254 @@
+      return;
+    }
+    const result = await fetchRecipes(searchQuery);
+    navigate(`/searchresults?query=${searchQuery}`);
```
update these navigations according to the changes route

shubhamships commented on gaurav/src/pages/recipe/SearchResults.tsx:
```diff
@@ -0,0 +1,164 @@
+
+  const query = new URLSearchParams(location.search).get("query");
+
+  const API_URL = `https://www.themealdb.com/api/json/v1/1/search.php?s=${query}`;
```
instead of declaring different API_URL variables in 3 places, declare at one and use it in all other places

shubhamships commented on gaurav/src/pages/recipe/SearchResults.tsx:
```diff
@@ -0,0 +1,164 @@
+    e.preventDefault();
+    const result = await fetchRecipes(searchQuery);
+    setSearchResults(result);
+    navigate(`/searchresults?query=${searchQuery}`);
```
update navigate link

gauravThakur-sgl commented:
@shubhamships Sir
Updated Changes :-

- Created `apiClient.tsx` where `baseURL: "https://www.themealdb.com/api/json/v1/1/"` and used it at all other place
- updated navigate link

shubhamships commented on gaurav/src/pages/recipe/RecipeDetail.tsx:
```diff
@@ -0,0 +1,102 @@
+  };
+  useEffect(() => {
+    setTimeout(() => {
+      setTimeOut(false);
```
setTimeOut is not a valid name as it is a reserved keyword

shubhamships commented on gaurav/src/pages/recipe/RecipeDetail.tsx:
```diff
@@ -0,0 +1,102 @@
+      ingredients.push(`${recipe[`strIngredient${i}`]} - ${recipe[`strMeasure${i}`]}`);
+    }
+  }
+  const steps = recipe.strInstructions.split(/(?:STEP\s*\d*[:.]?\s*)/i).filter((step) => step.trim() !== "");
```
what does this do?

gauravThakur-sgl replied:
this  splits the given api single response string into different steps
ex - `STEP 1\r\nTO MAKE SUSHI ROLLS: Pat out some rice. Lay a nori sheet on the mat, shiny-side down. Dip your hands in the vinegared water, then pat handfuls of rice on top in a 1cm thick layer, leaving the furthest edge from you clear.\r\n\r\nSTEP 2\r\nSpread over some Japanese mayonnaise. Use a spoon to spread out a thin layer of mayonnaise down the middle of the rice.` into separate steps

shubhamships commented on gaurav/src/pages/recipe/SavedRecipe.tsx:
```diff
@@ -0,0 +1,78 @@
+      <div className="h-screen text-2xl font-bold bg-recipebg text-white flex items-center justify-center">
+        Your List is Empty
+      </div>
+    );
```
use it below

gauravThakur-sgl commented:
Updated Changes :- 

- used `isTimeOut` instead of `timeOut` and `setTimeOut`
- Used `<div className="h-screen text-2xl font-bold bg-recipebg text-white flex items-center justify-center">
        Your List is Empty
      </div>` inside the single Return element based on condition.

shubhamships commented on gaurav/src/pages/recipe/HomePage.tsx:
```diff
@@ -0,0 +1,253 @@
+      setRecipes(result);
+      const [categoriesData, areasData] = await Promise.all([fetchCategories(), fetchAreas()]);
+      setCategories(["all", ...categoriesData]);
+      setAreas(["all", ...areasData]);
```
incompatible data types beings added in categories and areas array....why are we adding all here?

gauravThakur-sgl replied:
@shubhamships Sir, the first element is set to `all` because it is selected by default. This ensures that the category and areas are displayed appropriately.
